linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,343 @@
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            if (ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care) != THREADAPI_OK)
+            {
+                LogError("Failure joining thread");
+            }
+        }
+
+        // Loop through
+        while (threadpool->task_list != NULL)
+        {
+            THREADPOOL_TASK* temp_item = threadpool->task_list;
+            threadpool->task_list = temp_item->next;
+            free(temp_item);
+        }
+
+        free(threadpool->thread_handle_list);
+
+        sem_destroy(&threadpool->semaphore);
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            {
+                srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+                if (threadpool->task_list == NULL)
+                {
+                    threadpool->task_list = task_item;
+                }
+                else
+                {
+                    threadpool->task_list_last->next = task_item;
+                }
+                threadpool->task_list_last = task_item;
+                (void)interlocked_increment(&threadpool->task_count);
+
+                srw_lock_release_exclusive(threadpool->srw_lock);
+            }
+
+            do
+            {
+                // Check to see if we should create another thread
+                int32_t current_count = interlocked_add(&threadpool->thread_count, 0);
+                if (current_count < threadpool->max_thread_count)
+                {
+                    if (interlocked_compare_exchange(&threadpool->thread_count, current_count+1, current_count) == current_count)
```

Comment:
likely missing a loop here, we make a decision based on an interlocked, then compare-exchange, but if that fails, then something happened and we should re-evaluate.